pagopa/io-app
# feat: [IOPID-315] LV at startup

## Short description
This PR, with some mocked clients, enables fast login functionality in startup saga.

## List of changes proposed in this pull request
- ts/components/cie/CieRequestAuthenticationOverlay.tsx: added selector for fast login ff.
- ts/features/fastLogin/backend/index.ts: added functions to perform a fast login.
- ts/features/fastLogin/backend/mockedClients.ts: added mocked clients to perform a fast login.
- ts/features/fastLogin/backend/mockedFunctionsAndTypes.ts: added mocked types to perform a fast login.
- ts/features/fastLogin/saga/tokenRefreshSaga.ts: added saga that perform a complete fast login, from getting Nonce to the actual fast login.
- ts/features/lollipop/hooks/useLollipopLoginSource.tsx: added selector for fast login ff.
- ts/features/lollipop/index.ts: added header for login type (Standard or FL).
- ts/features/lollipop/saga/index.ts: added function to get KeyInfo.
- [ts/features/lollipop/utils/__tests__/retryAndVerifyMismatch.test.ts, ts/features/lollipop/utils/__tests__/retryAndVerifyMissingSamlRequest.test.ts, ts/features/lollipop/utils/__tests__/retryAndVerifyMissingThumbprint.test.ts, ts/features/lollipop/utils/__tests__/retryAndVerifyRedirectNativeError.test.ts, ts/features/lollipop/utils/__tests__/retryAndVerifySuccess.test.ts]: Updated tests.
- ts/features/lollipop/utils/fetch.ts: added to lollipopFetch a maxRetries parameter.
- ts/features/lollipop/utils/login.ts: added parameter for Fast Login.
- ts/sagas/__tests__/initializeApplicationSaga.test.ts: updated and added test.
- ts/sagas/index.ts: added saga watch for fast login refresh token.
- ts/sagas/startup.ts: added check on 401 response from getSession.
- ts/screens/authentication/idpAuthSessionHandler.tsx: added fast login selector.
- ts/store/actions/authentication.ts: added action to refresh token for fast login.
- ts/store/reducers/authentication.ts: added reducer to set the new refreshed token.


## How to test
1) Run dev server and configure it for fast login. Set an expiration date for assertionRef too.
2) Run IO app and authenticate.
3) Close the app.
4) Reopen the app after the duration time set for the token in the dev server.
5) It should perform a fast login:
    - getSession : 401
    - getNonce: 200
    - fastLogin: 200
6) Close the app.
7) Reopen the app after the duration time set for the assertionRef in the dev server.
8) It should log you out.

To test the joint retry mechanism for getNonce and fastLogin, with a tool like ProxyMan, try to return error states for those calls

<details><summary>Dev server configuration exmpale</summary>
<p>

```json
{
  "profile": {
    "attrs": {
      "name": "Gian Maria"
    }
  },
  "features": {
    "lollipop": {
      "enabled": true,
      "assertionRefValidityMS": 120000
    },
    "fastLogin": {
      "sessionTTLinMS": 60000
    }
  }
}
```

</p>
</details>
